In [1]:
import matplotlib
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import pandas as pd
import numpy as np

data = pd.read_csv('fifa21_train.csv')

/Users/chrislacarte/opt/anaconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3444: DtypeWarning: Columns (70) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [2]:
data.shape

(11701, 101)

We convert column names to lwer case and eliminate spaces:

In [3]:
cols = []
for i in range(len(data.columns)):
    cols.append(data.columns[i].lower().replace(' ', '_'))
data.columns = cols

In [4]:
cols

['id',
 'name',
 'age',
 'nationality',
 'club',
 'bp',
 'position',
 'team_&_contract',
 'height',
 'weight',
 'foot',
 'growth',
 'joined',
 'loan_date_end',
 'value',
 'wage',
 'release_clause',
 'contract',
 'attacking',
 'crossing',
 'finishing',
 'heading_accuracy',
 'short_passing',
 'volleys',
 'skill',
 'dribbling',
 'curve',
 'fk_accuracy',
 'long_passing',
 'ball_control',
 'movement',
 'acceleration',
 'sprint_speed',
 'agility',
 'reactions',
 'balance',
 'power',
 'shot_power',
 'jumping',
 'stamina',
 'strength',
 'long_shots',
 'mentality',
 'aggression',
 'interceptions',
 'positioning',
 'vision',
 'penalties',
 'composure',
 'defending',
 'marking',
 'standing_tackle',
 'sliding_tackle',
 'goalkeeping',
 'gk_diving',
 'gk_handling',
 'gk_kicking',
 'gk_positioning',
 'gk_reflexes',
 'total_stats',
 'base_stats',
 'w/f',
 'sm',
 'a/w',
 'd/w',
 'ir',
 'pac',
 'sho',
 'pas',
 'dri',
 'defe',
 'phy',
 'hits',
 'ls',
 'st',
 'rs',
 'lw',
 'lf',
 'cf',
 'rf',
 'rw',
 'lam

In [5]:
data

id            name  age  nationality                        club  \
0      184383       A. Pasche   26  Switzerland           FC Lausanne-Sport   
1      188044   Alan Carvalho   30     China PR     Beijing Sinobo Guoan FC   
2      184431     S. Giovinco   33        Italy                    Al Hilal   
3      233796        J. Evans   22        Wales                Swansea City   
4      234799      Y. Demoncy   23       France  US Orléans Loiret Football   
...       ...             ...  ...          ...                         ...   
11696  232504   B. Böðvarsson   25      Iceland       Jagiellonia Białystok   
11697  214680       G. Gallon   27       France                ESTAC Troyes   
11698  221489       J. Flores   22        Chile              CD Antofagasta   
11699  146717  Anderson Silva   26       Brazil                    Barnsley   
11700  236507     T. Conechny   22    Argentina            Portland Timbers   

        bp   position                               team_&_contract height  \
0       CM     CM CDM                 FC Lausanne-Sport 2015 ~ 2020   5'9"   
1       ST   ST LW LM  Beijing Sinobo Guoan FC Dec 31, 2020 On Loan   6'0"   
2      CAM     CAM CF                          Al Hilal 2019 ~ 2022   5'4"   
3      CDM     CDM CM                      Swansea City 2016 ~ 2021  5'10"   
4      CDM     CDM CM        US Orléans Loiret Football 2018 ~ 2021  5'11"   
...    ...        ...                                           ...    ...   
11696   LB         LB             Jagiellonia Białystok 2018 ~ 2021   6'1"   
11697   GK         GK                      ESTAC Troyes 2019 ~ 2022   6'1"   
11698   RM  LM CAM RM                    CD Antofagasta 2019 ~ 2024   5'6"   
11699   CM        NaN                                 Barnsley 2010   6'2"   
11700  CAM  ST CAM LM                  Portland Timbers 2019 ~ 2020   5'7"   

       weight  ...   cdm   rdm   rwb    lb   lcb    cb   rcb    rb    gk  ova  
0      161lbs  ...  59+1  59+1  59+1  58+1  54+1  54+1  54+1  58+1  15+1   64  
1      159lbs  ...  53+2  53+2  57+2  53+2  48+2  48+2  48+2  53+2  18+2   77  
2      134lbs  ...  56+2  56+2  59+2  53+2  41+2  41+2  41+2  53+2  12+2   80  
3      152lbs  ...  58+2  58+2  56+2  57+2  58+2  58+2  58+2  57+2  14+2   59  
4      150lbs  ...  64+2  64+2  64+2  63+2  61+2  61+2  61+2  63+2  15+2   65  
...       ...  ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...  
11696  168lbs  ...  60+2  60+2  63+2  63+2  61+2  61+2  61+2  63+2  16+2   65  
11697  174lbs  ...  26+2  26+2  25+2  24+2  26+2  26+2  26+2  24+2  69+2   70  
11698  143lbs  ...  44+2  44+2  49+2  45+2  35+2  35+2  35+2  45+2  17+2   67  
11699  179lbs  ...  68+0  68+0  66+0  64+0  60+0  60+0  60+0  64+0  25+0   68  
11700  143lbs  ...  43+2  43+2  46+2  44+2  36+2  36+2  36+2  44+2  15+2   64  

[11701 rows x 101 columns]

Colummns that we consider are not relevant for the OVA calculation:
id
name
nationality
club
bp (wee keep positon)
team & contract

In [6]:
data_num = data._get_numeric_data()
data_num

id  age  growth  attacking  crossing  finishing  heading_accuracy  \
0      184383   26       1        258        54         47                43   
1      188044   30       0        365        66         79                76   
2      184431   33       0        336        73         76                34   
3      233796   22      13        242        44         42                58   
4      234799   23       8        249        49         37                61   
...       ...  ...     ...        ...       ...        ...               ...   
11696  232504   25       5        224        62         24                51   
11697  214680   27       4         84        12         14                13   
11698  221489   22       8        286        64         66                51   
11699  146717   26       7        254        64         66                51   
11700  236507   22       9        289        43         64                65   

       short_passing  volleys  skill  ...  gk_positioning  gk_reflexes  \
0                 70     44.0    286  ...               9            6   
1                 68     76.0    375  ...               7           16   
2                 78     75.0    424  ...               3            3   
3                 62     36.0    259  ...               7           12   
4                 68     34.0    280  ...               5           15   
...              ...      ...    ...  ...             ...          ...   
11696             59     28.0    248  ...               6           14   
11697             29     16.0     96  ...              70           72   
11698             64     41.0    291  ...               7           16   
11699             73      NaN    284  ...              24           24   
11700             64     53.0    267  ...               8            8   

       total_stats  base_stats  pac  sho  pas  dri  phy  ova  
0             1682         357   69   51   63   63   60   64  
1             1961         412   83   75   68   82   71   77  
2             1925         404   80   77   78   86   56   80  
3             1527         329   57   44   54   57   60   59  
4             1664         360   66   44   60   64   66   65  
...            ...         ...  ...  ...  ...  ...  ...  ...  
11696         1589         347   75   28   56   59   69   65  
11697         1123         387   72   67   66   72   70   70  
11698         1603         337   76   60   60   72   43   67  
11699         1429         400   72   63   70   72   66   68  
11700         1607         336   79   62   56   67   48   64  

[11701 rows x 52 columns]

Check for NaN values: 

In [7]:
data_num.isna().sum()

id                    0
age                   0
growth                0
attacking             0
crossing              0
finishing             0
heading_accuracy      0
short_passing         0
volleys              41
skill                 0
dribbling             0
curve                41
fk_accuracy           0
long_passing          0
ball_control          0
movement              0
acceleration          0
sprint_speed          0
agility              41
reactions             0
balance              41
power                 0
shot_power            0
jumping              41
stamina               0
strength              0
long_shots            0
mentality             0
aggression            0
interceptions         7
positioning           7
vision               41
penalties             0
composure           279
defending             0
marking               0
standing_tackle       0
sliding_tackle       41
goalkeeping           0
gk_diving             0
gk_handling           0
gk_kicking      

In [8]:
nulls_df = pd.DataFrame(round(data_num.isna().sum()/len(data_num),4)*100)
nulls_df

0
id                0.00
age               0.00
growth            0.00
attacking         0.00
crossing          0.00
finishing         0.00
heading_accuracy  0.00
short_passing     0.00
volleys           0.35
skill             0.00
dribbling         0.00
curve             0.35
fk_accuracy       0.00
long_passing      0.00
ball_control      0.00
movement          0.00
acceleration      0.00
sprint_speed      0.00
agility           0.35
reactions         0.00
balance           0.35
power             0.00
shot_power        0.00
jumping           0.35
stamina           0.00
strength          0.00
long_shots        0.00
mentality         0.00
aggression        0.00
interceptions     0.06
positioning       0.06
vision            0.35
penalties         0.00
composure         2.38
defending         0.00
marking           0.00
standing_tackle   0.00
sliding_tackle    0.35
goalkeeping       0.00
gk_diving         0.00
gk_handling       0.00
gk_kicking        0.00
gk_positioning    0.00
gk_reflexes       0.00
total_stats       0.00
base_stats        0.00
pac               0.00
sho               0.00
pas               0.00
dri               0.00
phy               0.00
ova               0.00

As the 0.35 % NaN are not relevant, we can drop all ther NaN 
Also for composure which is 2.38%

In [9]:
data['volleys'].isna().sum()

41

In [10]:
data_num = data_num[data_num['volleys'].isna()==False]

In [11]:
data_num['volleys'].isna().sum()

0

In [12]:
nulls_df = pd.DataFrame(round(data_num.isna().sum()/len(data_num),4)*100)
nulls_df

0
id                0.00
age               0.00
growth            0.00
attacking         0.00
crossing          0.00
finishing         0.00
heading_accuracy  0.00
short_passing     0.00
volleys           0.00
skill             0.00
dribbling         0.00
curve             0.00
fk_accuracy       0.00
long_passing      0.00
ball_control      0.00
movement          0.00
acceleration      0.00
sprint_speed      0.00
agility           0.00
reactions         0.00
balance           0.00
power             0.00
shot_power        0.00
jumping           0.00
stamina           0.00
strength          0.00
long_shots        0.00
mentality         0.00
aggression        0.00
interceptions     0.00
positioning       0.00
vision            0.00
penalties         0.00
composure         2.04
defending         0.00
marking           0.00
standing_tackle   0.00
sliding_tackle    0.00
goalkeeping       0.00
gk_diving         0.00
gk_handling       0.00
gk_kicking        0.00
gk_positioning    0.00
gk_reflexes       0.00
total_stats       0.00
base_stats        0.00
pac               0.00
sho               0.00
pas               0.00
dri               0.00
phy               0.00
ova               0.00

The 'composure" column still has 2.04% of NaN , lets find out why:

In [13]:
data_num['composure'].isna().sum()

238

We decide to drop "composure" as it is not relevant for the total sample size

In [14]:
data_num = data_num[data_num['composure'].isna()==False]

In [15]:
nulls_df = pd.DataFrame(round(data_num.isna().sum()/len(data_num),4)*100)
nulls_df

0
id                0.0
age               0.0
growth            0.0
attacking         0.0
crossing          0.0
finishing         0.0
heading_accuracy  0.0
short_passing     0.0
volleys           0.0
skill             0.0
dribbling         0.0
curve             0.0
fk_accuracy       0.0
long_passing      0.0
ball_control      0.0
movement          0.0
acceleration      0.0
sprint_speed      0.0
agility           0.0
reactions         0.0
balance           0.0
power             0.0
shot_power        0.0
jumping           0.0
stamina           0.0
strength          0.0
long_shots        0.0
mentality         0.0
aggression        0.0
interceptions     0.0
positioning       0.0
vision            0.0
penalties         0.0
composure         0.0
defending         0.0
marking           0.0
standing_tackle   0.0
sliding_tackle    0.0
goalkeeping       0.0
gk_diving         0.0
gk_handling       0.0
gk_kicking        0.0
gk_positioning    0.0
gk_reflexes       0.0
total_stats       0.0
base_stats        0.0
pac               0.0
sho               0.0
pas               0.0
dri               0.0
phy               0.0
ova               0.0

Explore Numerical data

In [16]:
data_num


id  age  growth  attacking  crossing  finishing  heading_accuracy  \
0      184383   26       1        258        54         47                43   
1      188044   30       0        365        66         79                76   
2      184431   33       0        336        73         76                34   
3      233796   22      13        242        44         42                58   
4      234799   23       8        249        49         37                61   
...       ...  ...     ...        ...       ...        ...               ...   
11695  196848   29       0        290        36         72                65   
11696  232504   25       5        224        62         24                51   
11697  214680   27       4         84        12         14                13   
11698  221489   22       8        286        64         66                51   
11700  236507   22       9        289        43         64                65   

       short_passing  volleys  skill  ...  gk_positioning  gk_reflexes  \
0                 70     44.0    286  ...               9            6   
1                 68     76.0    375  ...               7           16   
2                 78     75.0    424  ...               3            3   
3                 62     36.0    259  ...               7           12   
4                 68     34.0    280  ...               5           15   
...              ...      ...    ...  ...             ...          ...   
11695             57     60.0    283  ...              13           14   
11696             59     28.0    248  ...               6           14   
11697             29     16.0     96  ...              70           72   
11698             64     41.0    291  ...               7           16   
11700             64     53.0    267  ...               8            8   

       total_stats  base_stats  pac  sho  pas  dri  phy  ova  
0             1682         357   69   51   63   63   60   64  
1             1961         412   83   75   68   82   71   77  
2             1925         404   80   77   78   86   56   80  
3             1527         329   57   44   54   57   60   59  
4             1664         360   66   44   60   64   66   65  
...            ...         ...  ...  ...  ...  ...  ...  ...  
11695         1638         337   66   68   47   66   57   65  
11696         1589         347   75   28   56   59   69   65  
11697         1123         387   72   67   66   72   70   70  
11698         1603         337   76   60   60   72   43   67  
11700         1607         336   79   62   56   67   48   64  

[11422 rows x 52 columns]

In [17]:
data_num.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11422 entries, 0 to 11700
Data columns (total 52 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                11422 non-null  int64  
 1   age               11422 non-null  int64  
 2   growth            11422 non-null  int64  
 3   attacking         11422 non-null  int64  
 4   crossing          11422 non-null  int64  
 5   finishing         11422 non-null  int64  
 6   heading_accuracy  11422 non-null  int64  
 7   short_passing     11422 non-null  int64  
 8   volleys           11422 non-null  float64
 9   skill             11422 non-null  int64  
 10  dribbling         11422 non-null  int64  
 11  curve             11422 non-null  float64
 12  fk_accuracy       11422 non-null  int64  
 13  long_passing      11422 non-null  int64  
 14  ball_control      11422 non-null  int64  
 15  movement          11422 non-null  int64  
 16  acceleration      11422 non-null  int64 

In [18]:
data_num.describe(include="all").T

count           mean           std    min        25%  \
id                11422.0  221862.372527  32191.114029   41.0  205360.25   
age               11422.0      25.155577      4.872713   16.0      21.00   
growth            11422.0       5.593241      5.824869    0.0       0.00   
attacking         11422.0     258.114253     72.216611   42.0     231.25   
crossing          11422.0      51.499037     17.853203    6.0      41.00   
finishing         11422.0      47.941779     19.385843    3.0      33.00   
heading_accuracy  11422.0      53.377079     16.918882    5.0      45.00   
short_passing     11422.0      60.363947     13.939404   11.0      56.00   
volleys           11422.0      44.932411     17.682967    4.0      32.00   
skill             11422.0     266.132726     76.793575   43.0     232.00   
dribbling         11422.0      57.805901     18.062550    5.0      53.00   
curve             11422.0      49.497286     18.140603    4.0      37.00   
fk_accuracy       11422.0      44.156190     17.391467    5.0      32.00   
long_passing      11422.0      54.139117     14.875551    9.0      45.00   
ball_control      11422.0      60.534232     15.864516    5.0      57.00   
movement          11422.0     323.169060     54.975768  113.0     295.00   
acceleration      11422.0      65.494047     14.726676   13.0      58.00   
sprint_speed      11422.0      65.519436     14.496461   11.0      59.00   
agility           11422.0      64.621432     14.490637   14.0      57.00   
reactions         11422.0      62.823936      8.982013   24.0      57.00   
balance           11422.0      64.710208     14.144392   17.0      57.00   
power             11422.0     302.216337     49.607925  133.0     271.00   
shot_power        11422.0      59.655227     13.414549   12.0      50.00   
jumping           11422.0      65.046927     12.013735   25.0      58.00   
stamina           11422.0      63.334267     15.324296   12.0      56.00   
strength          11422.0      65.153826     12.727787   20.0      58.00   
long_shots        11422.0      49.026090     19.045676    4.0      35.00   
mentality         11422.0     261.386185     62.814132   55.0     234.00   
aggression        11422.0      56.867011     17.063100    9.0      45.00   
interceptions     11422.0      46.730870     20.790790    5.0      26.00   
positioning       11422.0      52.372089     19.047291    2.0      43.00   
vision            11422.0      55.364472     13.652586   10.0      47.00   
penalties         11422.0      50.051742     15.597743    8.0      40.00   
composure         11422.0      59.947732     11.769872   12.0      53.00   
defending         11422.0     140.945018     61.494173   20.0      83.00   
marking           11422.0      47.021712     20.133715    3.0      29.00   
standing_tackle   11422.0      48.066276     21.446926    6.0      28.00   
sliding_tackle    11422.0      45.857030     21.198218    6.0      24.00   
goalkeeping       11422.0      77.767554     81.319760   12.0      48.00   
gk_diving         11422.0      15.634215     16.867775    2.0       8.00   
gk_handling       11422.0      15.500700     16.155198    2.0       8.00   
gk_kicking        11422.0      15.355017     15.810872    2.0       8.00   
gk_positioning    11422.0      15.525214     16.403054    2.0       8.00   
gk_reflexes       11422.0      15.752408     17.176922    2.0       8.00   
total_stats       11422.0    1629.731133    260.649387  731.0    1491.00   
base_stats        11422.0     360.993434     40.283405  228.0     333.00   
pac               11422.0      68.179391     11.182793   26.0      62.00   
sho               11422.0      54.942479     13.830244   17.0      46.00   
pas               11422.0      58.813956     10.129438   25.0      52.00   
dri               11422.0      64.193486      9.800793   28.0      59.00   
phy               11422.0      64.827526      9.741849   27.0      59.00   
ova               11422.0      66.876642      6.845047   45.0      6

In [19]:
# for column in data_num.columns:
#     sns.distplot(data_num[column])
#     plt.show()

In [20]:
# correlations_matrix = data_num.corr()
# correlations_matrix

In [21]:
# sns.heatmap(correlations_matrix, annot=True)
# plt.show()

After revising the data and studying the documentation for the way "ova" is calculated, we decide the relevant columns are as follows:
    

In [22]:
list_dummies = ['ls','st','rs','lw','lf','cf','rf','rw','lam','cam','ram','lm','lcm','cm','rcm','rm','lwb','ldm','cdm','rdm','rwb','lb','lcb','cb','rcb','rb','gk',]

We standardize the way some of them represetn the data (e.g 59+1 = 59 ) and we eliminate the +1s as is just a booster that does not happen every observation:

In [23]:
for col in data[list_dummies]:
    data[col] = data[col].str.split("+", n = 1, expand = True)[0]
data

id            name  age  nationality                        club  \
0      184383       A. Pasche   26  Switzerland           FC Lausanne-Sport   
1      188044   Alan Carvalho   30     China PR     Beijing Sinobo Guoan FC   
2      184431     S. Giovinco   33        Italy                    Al Hilal   
3      233796        J. Evans   22        Wales                Swansea City   
4      234799      Y. Demoncy   23       France  US Orléans Loiret Football   
...       ...             ...  ...          ...                         ...   
11696  232504   B. Böðvarsson   25      Iceland       Jagiellonia Białystok   
11697  214680       G. Gallon   27       France                ESTAC Troyes   
11698  221489       J. Flores   22        Chile              CD Antofagasta   
11699  146717  Anderson Silva   26       Brazil                    Barnsley   
11700  236507     T. Conechny   22    Argentina            Portland Timbers   

        bp   position                               team_&_contract height  \
0       CM     CM CDM                 FC Lausanne-Sport 2015 ~ 2020   5'9"   
1       ST   ST LW LM  Beijing Sinobo Guoan FC Dec 31, 2020 On Loan   6'0"   
2      CAM     CAM CF                          Al Hilal 2019 ~ 2022   5'4"   
3      CDM     CDM CM                      Swansea City 2016 ~ 2021  5'10"   
4      CDM     CDM CM        US Orléans Loiret Football 2018 ~ 2021  5'11"   
...    ...        ...                                           ...    ...   
11696   LB         LB             Jagiellonia Białystok 2018 ~ 2021   6'1"   
11697   GK         GK                      ESTAC Troyes 2019 ~ 2022   6'1"   
11698   RM  LM CAM RM                    CD Antofagasta 2019 ~ 2024   5'6"   
11699   CM        NaN                                 Barnsley 2010   6'2"   
11700  CAM  ST CAM LM                  Portland Timbers 2019 ~ 2020   5'7"   

       weight  ... cdm  rdm rwb  lb lcb  cb rcb  rb  gk  ova  
0      161lbs  ...  59   59  59  58  54  54  54  58  15   64  
1      159lbs  ...  53   53  57  53  48  48  48  53  18   77  
2      134lbs  ...  56   56  59  53  41  41  41  53  12   80  
3      152lbs  ...  58   58  56  57  58  58  58  57  14   59  
4      150lbs  ...  64   64  64  63  61  61  61  63  15   65  
...       ...  ...  ..  ...  ..  ..  ..  ..  ..  ..  ..  ...  
11696  168lbs  ...  60   60  63  63  61  61  61  63  16   65  
11697  174lbs  ...  26   26  25  24  26  26  26  24  69   70  
11698  143lbs  ...  44   44  49  45  35  35  35  45  17   67  
11699  179lbs  ...  68   68  66  64  60  60  60  64  25   68  
11700  143lbs  ...  43   43  46  44  36  36  36  44  15   64  

[11701 rows x 101 columns]

We do the same wiht ['w/f','sm','ir'] as they have a * and an int in the value

In [24]:
dummies_2 = ['w/f','sm','ir']

for col in data[dummies_2]:
    data[col] = data[col].str.split(" ", n = 1, expand = True)[0]
data['w/f']



0        4
1        3
2        4
3        2
4        2
        ..
11696    3
11697    2
11698    3
11699    3
11700    3
Name: w/f, Length: 11701, dtype: object

we drop nationality , wage  as we consider they are nto relevant for the player avo

In [25]:
valid_columns = ['age','bp','height','weight','foot','growth','value','attacking','skill','movement','power','mentality','defending','goalkeeping','base_stats','w/f','sm','a/w','d/w','ir','pac','sho','pas','dri','defe','phy','hits','ls','st','rs','lw','lf','cf','rf','rw','lam','cam','ram','lm','lcm','cm','rcm','rm','lwb','ldm','cdm','rdm','rwb','lb','lcb','cb','rcb','rb','gk']

These are our selected columns:

In [26]:
data[valid_columns]

age   bp height  weight   foot  growth  value  attacking  skill  \
0       26   CM   5'9"  161lbs  Right       1  €525K        258    286   
1       30   ST   6'0"  159lbs  Right       0  €8.5M        365    375   
2       33  CAM   5'4"  134lbs  Right       0    €9M        336    424   
3       22  CDM  5'10"  152lbs  Right      13  €275K        242    259   
4       23  CDM  5'11"  150lbs  Right       8  €725K        249    280   
...    ...  ...    ...     ...    ...     ...    ...        ...    ...   
11696   25   LB   6'1"  168lbs   Left       5  €625K        224    248   
11697   27   GK   6'1"  174lbs  Right       4  €1.6M         84     96   
11698   22   RM   5'6"  143lbs  Right       8  €1.1M        286    291   
11699   26   CM   6'2"  179lbs  Right       7     €0        254    284   
11700   22  CAM   5'7"  143lbs   Left       9  €700K        289    267   

       movement  ...  ldm  cdm  rdm  rwb  lb lcb  cb rcb  rb  gk  
0           346  ...   59   59   59   59  58  54  54  54  58  15  
1           404  ...   53   53   53   57  53  48  48  48  53  18  
2           424  ...   56   56   56   59  53  41  41  41  53  12  
3           282  ...   58   58   58   56  57  58  58  58  57  14  
4           324  ...   64   64   64   64  63  61  61  61  63  15  
...         ...  ...  ...  ...  ...  ...  ..  ..  ..  ..  ..  ..  
11696       333  ...   60   60   60   63  63  61  61  61  63  16  
11697       237  ...   26   26   26   25  24  26  26  26  24  69  
11698       388  ...   44   44   44   49  45  35  35  35  45  17  
11699       198  ...   68   68   68   66  64  60  60  60  64  25  
11700       391  ...   43   43   43   46  44  36  36  36  44  15  

[11701 rows x 54 columns]

In [27]:
valid_columns

['age',
 'bp',
 'height',
 'weight',
 'foot',
 'growth',
 'value',
 'attacking',
 'skill',
 'movement',
 'power',
 'mentality',
 'defending',
 'goalkeeping',
 'base_stats',
 'w/f',
 'sm',
 'a/w',
 'd/w',
 'ir',
 'pac',
 'sho',
 'pas',
 'dri',
 'defe',
 'phy',
 'hits',
 'ls',
 'st',
 'rs',
 'lw',
 'lf',
 'cf',
 'rf',
 'rw',
 'lam',
 'cam',
 'ram',
 'lm',
 'lcm',
 'cm',
 'rcm',
 'rm',
 'lwb',
 'ldm',
 'cdm',
 'rdm',
 'rwb',
 'lb',
 'lcb',
 'cb',
 'rcb',
 'rb',
 'gk']

In [28]:
data.dtypes

id              int64
name           object
age             int64
nationality    object
club           object
                ...  
cb             object
rcb            object
rb             object
gk             object
ova             int64
Length: 101, dtype: object

DATA TRANSFORMATION: 

In [29]:
data['foot']

0        Right
1        Right
2        Right
3        Right
4        Right
         ...  
11696     Left
11697    Right
11698    Right
11699    Right
11700     Left
Name: foot, Length: 11701, dtype: object

In [30]:
from sklearn.preprocessing import LabelEncoder
label_encoded = LabelEncoder().fit(data['foot']).transform(data['foot']) # ordered wrt value counts
label_encoded = pd.DataFrame(label_encoded)
display(label_encoded.head(20))
# label_encoded['foot'].value_counts()
data['foot']= label_encoded

0
0   1
1   1
2   1
3   1
4   1
5   1
6   1
7   0
8   1
9   0
10  1
11  1
12  1
13  1
14  1
15  0
16  1
17  1
18  1
19  0

In [31]:
data[valid_columns]

age   bp height  weight  foot  growth  value  attacking  skill  \
0       26   CM   5'9"  161lbs     1       1  €525K        258    286   
1       30   ST   6'0"  159lbs     1       0  €8.5M        365    375   
2       33  CAM   5'4"  134lbs     1       0    €9M        336    424   
3       22  CDM  5'10"  152lbs     1      13  €275K        242    259   
4       23  CDM  5'11"  150lbs     1       8  €725K        249    280   
...    ...  ...    ...     ...   ...     ...    ...        ...    ...   
11696   25   LB   6'1"  168lbs     0       5  €625K        224    248   
11697   27   GK   6'1"  174lbs     1       4  €1.6M         84     96   
11698   22   RM   5'6"  143lbs     1       8  €1.1M        286    291   
11699   26   CM   6'2"  179lbs     1       7     €0        254    284   
11700   22  CAM   5'7"  143lbs     0       9  €700K        289    267   

       movement  ...  ldm  cdm  rdm  rwb  lb lcb  cb rcb  rb  gk  
0           346  ...   59   59   59   59  58  54  54  54  58  15  
1           404  ...   53   53   53   57  53  48  48  48  53  18  
2           424  ...   56   56   56   59  53  41  41  41  53  12  
3           282  ...   58   58   58   56  57  58  58  58  57  14  
4           324  ...   64   64   64   64  63  61  61  61  63  15  
...         ...  ...  ...  ...  ...  ...  ..  ..  ..  ..  ..  ..  
11696       333  ...   60   60   60   63  63  61  61  61  63  16  
11697       237  ...   26   26   26   25  24  26  26  26  24  69  
11698       388  ...   44   44   44   49  45  35  35  35  45  17  
11699       198  ...   68   68   68   66  64  60  60  60  64  25  
11700       391  ...   43   43   43   46  44  36  36  36  44  15  

[11701 rows x 54 columns]

In [32]:
# for col in data['weight']:
#     data['weight'] = data['weight'].str.split("lbs", n = 1, expand = True)[0]
  
# data['weight']

In [33]:
#strip value and wage from the € symbol and the K, M, B
data['value'] = data['value'].str.replace("€", "").str.replace("K","000").str.replace("M","000000").str.replace("B","000000000")
data['value']

0           525000
1        8.5000000
2          9000000
3           275000
4           725000
           ...    
11696       625000
11697    1.6000000
11698    1.1000000
11699            0
11700       700000
Name: value, Length: 11701, dtype: object

In [34]:
#Ecncode A/W and D/W with LabelEncoder 
#A/W
df3 = data['a/w']
label_encoded = LabelEncoder().fit(df3).transform(df3) # ordered wrt value counts
label_encoded = pd.DataFrame(label_encoded)
display(label_encoded.head(20))
# label_encoded['foot'].value_counts()
data['a/w']= label_encoded

#D/W
df4 = data['d/w']
label_encoded = LabelEncoder().fit(df4).transform(df4) # ordered wrt value counts
label_encoded = pd.DataFrame(label_encoded)
display(label_encoded.head(20))
# label_encoded['foot'].value_counts()
data['d/w'] = label_encoded


0
0   0
1   0
2   0
3   2
4   1
5   0
6   2
7   2
8   2
9   2
10  2
11  0
12  2
13  2
14  2
15  0
16  2
17  2
18  2
19  2

0
0   2
1   1
2   2
3   2
4   2
5   1
6   0
7   2
8   2
9   2
10  2
11  1
12  2
13  2
14  2
15  1
16  2
17  2
18  1
19  2

 We need to transform feet, inches  into cm

In [46]:
dummy = data.head(5)
dummy

id           name  age  nationality                        club   bp  \
0  184383      A. Pasche   26  Switzerland           FC Lausanne-Sport   CM   
1  188044  Alan Carvalho   30     China PR     Beijing Sinobo Guoan FC   ST   
2  184431    S. Giovinco   33        Italy                    Al Hilal  CAM   
3  233796       J. Evans   22        Wales                Swansea City  CDM   
4  234799     Y. Demoncy   23       France  US Orléans Loiret Football  CDM   

   position                               team_&_contract height  weight  ...  \
0    CM CDM                 FC Lausanne-Sport 2015 ~ 2020   5'9"  161lbs  ...   
1  ST LW LM  Beijing Sinobo Guoan FC Dec 31, 2020 On Loan   6'0"  159lbs  ...   
2    CAM CF                          Al Hilal 2019 ~ 2022   5'4"  134lbs  ...   
3    CDM CM                      Swansea City 2016 ~ 2021  5'10"  152lbs  ...   
4    CDM CM        US Orléans Loiret Football 2018 ~ 2021  5'11"  150lbs  ...   

   cdm  rdm rwb  lb lcb  cb rcb  rb  gk  ova  
0   59   59  59  58  54  54  54  58  15   64  
1   53   53  57  53  48  48  48  53  18   77  
2   56   56  59  53  41  41  41  53  12   80  
3   58   58  56  57  58  58  58  57  14   59  
4   64   64  64  63  61  61  61  63  15   65  

[5 rows x 101 columns]

In [47]:
# h_inch += h_ft * 12
# h_cm = round(h_inch * 2.54, 1)


# \"



for col in dummy['height']:
    data_height_ft = dummy['height'].str.split("\'", n = 1, expand = True)[0]
    data_height_in_1 = dummy['height'].str.split("\'", n = 1, expand = True)[1]    
    for i in data_height_in_1:
        data_height_in_2 = i.rstrip("\"")
    data_height_ft = pd.to_numeric(data_height_ft)
    data_height_in = pd.to_numeric(data_height_in_2)
print(data_height_ft)
print(data_height_in)



dummy2 = (data_height_ft * 30.48)  + (data_height_in * 2.54)
dummy2

0    5
1    6
2    5
3    5
4    5
Name: 0, dtype: int64
11


0    180.34
1    210.82
2    180.34
3    180.34
4    180.34
Name: 0, dtype: float64

In [ ]:
for col in data['height']:
    data_height_ft = data['height'].str.split("\'", n = 1, expand = True)[0]
    data_height_in_1 = data['height'].str.split("\'", n = 1, expand = True)[1]    
    for i in data_height_in_1:
        data_height_in_2 = i.rstrip("\"")
    data_height_ft = pd.to_numeric(data_height_ft)
    data_height_in = pd.to_numeric(data_height_in_2)
print(data_height_ft)
print(data_height_in)



data['height'] = (data_height_ft * 30.48)  + (data_height_in * 2.54)
data['height']

In [60]:
# def inch_to_cm(inch):
#     return inch * 2.54

# def feet_to_cm(feet):
#     return feet * 30.48 



In [48]:
for item in data['height']:
    data[item] = data[item].str.split(("'"),n = 1, expand = True)[0]
data


KeyError: '5\'9"'

In [46]:
def clean_fifa21(df):
    cols = []
    for i in range(len(data.columns)):
        cols.append(data.columns[i].lower().replace(' ', '_'))
    data.columns = cols
    
    data_num = data._get_numeric_data()
    data_num = data_num[data_num['volleys'].isna()==False]
    data_num = data_num[data_num['composure'].isna()==False]
    
    list_dummies = ['ls','st','rs','lw','lf','cf','rf','rw','lam','cam','ram','lm','lcm','cm','rcm','rm','lwb','ldm','cdm','rdm','rwb','lb','lcb','cb','rcb','rb','gk',]
    for col in data[list_dummies]:
        data[col] = data[col].str.split("+", n = 1, expand = True)[0]
    
    dummies_2 = ['w/f','sm','ir']
    for col in data[dummies_2]:
        data[col] = data[col].str.split(" ", n = 1, expand = True)[0]
    
    valid_columns = ['age','bp','height','weight','foot','growth','value','attacking','skill','movement','power','mentality','defending','goalkeeping','base_stats','w/f','sm','a/w','d/w','ir','pac','sho','pas','dri','defe','phy','hits','ls','st','rs','lw','lf','cf','rf','rw','lam','cam','ram','lm','lcm','cm','rcm','rm','lwb','ldm','cdm','rdm','rwb','lb','lcb','cb','rcb','rb','gk']
    

    from sklearn.preprocessing import LabelEncoder
    label_encoded = LabelEncoder().fit(data['foot']).transform(data['foot']) # ordered wrt value counts
    label_encoded = pd.DataFrame(label_encoded)
    display(label_encoded.head(20))
    # label_encoded['foot'].value_counts()
    data['foot']= label_encoded
    
    for col in data['weight']:
    data['weight'] = data['weight'].str.split("lbs", n = 1, expand = True)[0]
    
    #A/W
    label_encoded = LabelEncoder().fit(data['a/w']).transform(data['a/w']) # ordered wrt value counts
    label_encoded = pd.DataFrame(label_encoded)
    display(label_encoded.head(20))
    # label_encoded['foot'].value_counts()
    data['a/w']= label_encoded

    #D/W
    df4 = data['d/w']
    label_encoded = LabelEncoder().fit(df4).transform(df4) # ordered wrt value counts
    label_encoded = pd.DataFrame(label_encoded)
    display(label_encoded.head(20))
    # label_encoded['foot'].value_counts()
    data['d/w'] = label_encoded
    
    data['value'] = data['value'].str.replace("€", "").str.replace("K","000").str.replace("M","000000").str.replace("B","000000000")
    
    